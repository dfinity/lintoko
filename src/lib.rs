use std::{collections::HashSet, fs, io::Write, path::Path};

use anyhow::{Context, Result, anyhow};
use miette::{LabeledSpan, NamedSource, Report, Severity, miette};
use serde::Deserialize;
use tracing::debug;
use tree_sitter::{Node, Parser, Query, QueryCursor, Range, StreamingIterator};

#[derive(Debug, Deserialize)]
pub struct Rule {
    name: String,
    description: String,
    query: String,
}

pub fn default_rules() -> Vec<Rule> {
    vec![
        toml::from_str(include_str!("../default-rules/no-flexible.toml"))
            .expect("Failed to parse no-flexible rule"),
        toml::from_str(include_str!("../default-rules/no-stable.toml"))
            .expect("Failed to parse no-stable rule"),
        toml::from_str(include_str!("../default-rules/only-persistent-actor.toml"))
            .expect("Failed to parse only-persistent-actor rule"),
    ]
}

pub fn load_rule_from_file(path: &Path) -> Result<Rule> {
    let content = std::fs::read_to_string(path)?;
    let rule = toml::from_str(&content)?;
    Ok(rule)
}

pub fn load_rules_from_directory(dir: &Path) -> Result<Vec<Rule>> {
    let mut rules = vec![];
    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();
        if path.is_file() && path.extension().unwrap_or_default() == "toml" {
            debug!("Parsing extra rule at: {}", path.display());
            let rule = load_rule_from_file(&path)
                .with_context(|| anyhow!("Failed to parse rule from: '{}'", path.display()))?;
            rules.push(rule)
        }
    }
    Ok(rules)
}

fn apply_rule(rule: &Rule, tree: Node, input: &str) -> Result<Vec<Report>> {
    let query = Query::new(&tree_sitter_motoko::LANGUAGE.into(), &rule.query)
        .with_context(|| format!("Failed to create query for rule '{}'", rule.name))?;
    let error_capture_index = query.capture_index_for_name("error").with_context(|| {
        anyhow!(
            "Expected query to contain `@error` captures:\n{}",
            rule.query
        )
    })?;
    let filter_capture_index = query.capture_index_for_name("filter");

    let mut cursor = QueryCursor::new();
    let mut matches = cursor.matches(&query, tree, input.as_bytes());
    let mut filtered: HashSet<Range> = HashSet::new();
    let mut errors: Vec<Range> = Vec::new();
    while let Some(m) = matches.next() {
        for error_node in m.nodes_for_capture_index(error_capture_index) {
            errors.push(error_node.range());
        }
        if let Some(filter_capture_index) = filter_capture_index {
            for filter_node in m.nodes_for_capture_index(filter_capture_index) {
                filtered.insert(filter_node.range());
            }
        }
    }
    let diagnostics = errors
        .into_iter()
        .filter(|range| !filtered.contains(range))
        .map(|range| {
            miette!(
                severity = Severity::Error,
                labels = vec![LabeledSpan::new_primary_with_span(
                    Some(rule.description.clone()),
                    (range.start_byte, range.end_byte - range.start_byte)
                )],
                "[ERROR]: {}",
                rule.name
            )
        })
        .collect();
    Ok(diagnostics)
}

pub fn lint_file(path: &str, input: &str, rules: Vec<Rule>, mut out: impl Write) -> Result<usize> {
    let mut parser = Parser::new();
    parser
        .set_language(&tree_sitter_motoko::LANGUAGE.into())
        .expect("Error loading Motoko grammar");
    let tree = parser.parse(input.as_bytes(), None).unwrap();
    let mut diagnostics = Vec::new();
    for rule in &rules {
        diagnostics.extend(apply_rule(rule, tree.root_node(), input)?);
    }
    let count = diagnostics.len();
    diagnostics.sort_by_key(|d| d.labels().unwrap().next().unwrap().offset());
    for diagnostic in diagnostics {
        let pretty = diagnostic.with_source_code(NamedSource::new(path, input.to_string()));
        writeln!(&mut out, "{pretty:?}")?
    }
    Ok(count)
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn no_rules() {
        let mut out: Vec<u8> = vec![];
        let err_count = lint_file(
            "<input_path>",
            include_str!("../test-data.mo"),
            vec![],
            &mut out,
        )
        .unwrap();
        assert_eq!(err_count, 0);
        assert_eq!(str::from_utf8(&out).unwrap(), "");
    }

    #[test]
    fn it_lints_default_rules() {
        let mut out: Vec<u8> = vec![];
        unsafe {
            std::env::set_var("NO_COLOR", "1");
        }
        let rules = default_rules();
        let _err_count = lint_file(
            "<input_path>",
            include_str!("../test-data.mo"),
            rules,
            &mut out,
        )
        .unwrap();
        let lint_output = str::from_utf8(&out).unwrap();
        insta::assert_snapshot!(lint_output);
    }

    #[test]
    fn it_lints_custom_rules() {
        let mut out: Vec<u8> = vec![];
        unsafe {
            std::env::set_var("NO_COLOR", "1");
        }
        let mut rules = default_rules();
        rules.extend(load_rules_from_directory(Path::new("custom-rules")).unwrap());
        let _err_count = lint_file(
            "<input_path>",
            include_str!("../test-data.mo"),
            rules,
            &mut out,
        )
        .unwrap();
        let lint_output = str::from_utf8(&out).unwrap();
        insta::assert_snapshot!(lint_output);
    }
}
