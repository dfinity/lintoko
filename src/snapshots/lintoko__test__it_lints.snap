---
source: src/lib.rs
expression: lint_output
---
  × [ERROR]: no-result
   ╭─[<input_path>:1:1]
 1 │ import Result "mo:base/Result";
   · ───────────────┬──────────────
   ·                ╰── Do not use the Result pattern. Use Runtime.trap() instead.
 2 │ import OrderedMap "mo:base/OrderedMap";
   ╰────

  × [ERROR]: shared-caller
   ╭─[<input_path>:5:8]
 4 │ 
 5 │ shared (msg) actor class() {
   ·        ──┬──
   ·          ╰── Do not use `shared(msg)`. Use `shared({ caller})` instead.
 6 │   func letElse() {
   ╰────

  × [ERROR]: only-persistent-actor
   ╭─[<input_path>:5:14]
 4 │ 
 5 │ shared (msg) actor class() {
   ·              ──┬──
   ·                ╰── Only use persistent actors. Change `actor` to `persistent actor`.
 6 │   func letElse() {
   ╰────

  × [ERROR]: no-let-else
   ╭─[<input_path>:7:5]
 6 │   func letElse() {
 7 │     let ?x = null else { return };
   ·     ──────────────┬──────────────
   ·                   ╰── Do not use let-else. Use a switch instead.
 8 │   };
   ╰────

  × [ERROR]: no-flexible
    ╭─[<input_path>:10:3]
  9 │ 
 10 │   flexible let flexibleLet = 42;
    ·   ────┬───
    ·       ╰── Do not use the `flexible` keyword. Use `transient` instead.
 11 │   stable let stableLet = 42;
    ╰────

  × [ERROR]: no-stable
    ╭─[<input_path>:11:3]
 10 │   flexible let flexibleLet = 42;
 11 │   stable let stableLet = 42;
    ·   ───┬──
    ·      ╰── Do not use the `stable` keyword. It's implicit when using persistent actors.
 12 │ 
    ╰────

  × [ERROR]: shared-caller
    ╭─[<input_path>:13:16]
 12 │ 
 13 │   public shared(msg) func sharedCaller() {};
    ·                ──┬──
    ·                  ╰── Do not use `shared(msg)`. Use `shared({ caller})` instead.
 14 │   public shared msg func sharedCallerVar() {};
    ╰────

  × [ERROR]: shared-caller
    ╭─[<input_path>:14:17]
 13 │   public shared(msg) func sharedCaller() {};
 14 │   public shared msg func sharedCallerVar() {};
    ·                 ─┬─
    ·                  ╰── Do not use `shared(msg)`. Use `shared({ caller})` instead.
 15 │   public shared ({ caller }) func sharedCallerPat() {};
    ╰────

  × [ERROR]: transient-usage
    ╭─[<input_path>:19:3]
 18 │   transient let transientMap = OrderedMap.Make<Nat>(Nat.compare);
 19 │   transient let transientMap = OrderedMap.Make.Well<Nat>(Nat.compare);
    ·   ─────────────────────────────────┬─────────────────────────────────
    ·                                    ╰── Only use `transient` for OrderedMap.Make, OrderedSet.Make, and Random.Finite
 20 │ 
    ╰────

  × [ERROR]: assign-plus
    ╭─[<input_path>:21:3]
 20 │ 
 21 │   x := 1 + x;
    ·   ─────┬────
    ·        ╰── Use `+=` instead of reassigning the variable
 22 │   x := x - 1;
    ╰────

  × [ERROR]: assign-minus
    ╭─[<input_path>:22:3]
 21 │   x := 1 + x;
 22 │   x := x - 1;
    ·   ─────┬────
    ·        ╰── Use `-=` instead of reassigning the variable
 23 │   x := 1 - x;
    ╰────
